### DISCUSSION TITLE
Can GRDB safely be used from multiple (extension) targets in a shared app group?

### DISCUSSION THREAD

**Discussion by hannesoid** [groue/GRDB.swift#1343] [Q&A] [answered]
Let's say we have 
- an iOS app with a main app target and a widget extension (or more extensions).
- the app and the extension both access the same SQLite file in a shared [app group](https://developer.apple.com/documentation/xcode/configuring-app-groups/) folder.

Usually we need [NSFileCoordinator](https://developer.apple.com/documentation/foundation/nsfilecoordinator) to coordinate access, but I read somewhere that SQLite has some multi-process protections when reading or using transactions.

As I couldn't find specific info in the documentation - does SQLite itself or GRDB take care of coordinating access  in such a scenario, or do we still need to use `NFileCoordinator`?

**Comment by groue** [comment, marked as answer]
Hello @hannesoid, please check the [Sharing a Database](https://swiftpackageindex.com/groue/grdb.swift/v6.8.0/documentation/grdb/databasesharing) guide, full of `NSFileCoordinator` and other tips and techniques.

Don't miss the initial advice of the guide:

> Preventing errors that may happen due to database sharing is difficult. It is extremely difficult on iOS. And it is almost impossible to test.
> Always consider sharing plain files, or any other inter-process communication technique, before sharing an SQLite database.

**Reply by hannesoid** [reply]
ah great, thanks for pointing me to it!

**Reply by groue** [reply]
I guess I need to add a link to the main README ;-)

**Reply by groue** [reply]
> I guess I need to add a link to the main README ;-)

The link was already there, in the main [README](https://github.com/groue/GRDB.swift/tree/development):

> [Sharing a Database](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/databasesharing): Recommendations for App Group Containers and sandboxed macOS apps.

Maybe you could not find it because it does not use the keywords would were expecting - any suggestion that would improve the copy is welcomed!

**Reply by hannesoid** [reply]
Right, I'm not sure why I didn't see it, It's hard to reconstruct, I suppose I was searching for keywords like 

- app extensions
- extensions
- multiple
- multiple targets
- NSFileCoordinator
- file coordination
- coordinated
- file access

Perhaps extending that sentence to cover some of that would have made it a direct hit for me, something like: 

> [Sharing a Database](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/databasesharing): Recommendations for sandboxed macOS apps and App Group Containers, using a single database from multiple app extensions, multiple targets or apps, file coordination / `NSFileCoordinator`.

**Reply by groue** [reply]
Thank you for your generous and helpful retrospection 👍

`NSFileCoordinator` is a little bit low-level and overly specific, but I like "file coordination" which has an excellent [information scent](https://www.nngroup.com/articles/information-scent/)!

> [Sharing a Database](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/databasesharing): How to share an SQLite database between multiple processes - recommendations for App Group containers, App Extensions, App Sandbox, and file coordination.

Yeah, name-dropping of Apple technologies is exactly what we need here. "Does GRDB support App Extensions? Can I store my database in an app group container?"

I'll use the same exact sentence in the article header.

**Reply by hannesoid** [reply]
You're very welcome!
Sounds exactly right and nicely findable, thanks for taking so much care to improve the documentation.
(the documentation is overall outstandingly good and complete by the way, I love it, great work)

**Reply by groue** [reply]
You're welcome and thank you very much!
